src/fs.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018-2024 the Deno authors. MIT license.
+
+use std::io::Error;
+use std::io::ErrorKind;
+use std::io::Write;
+use std::path::Path;
+use std::path::PathBuf;
+
+use sys_traits::FsCanonicalize;
+use sys_traits::FsCreateDirAll;
+use sys_traits::FsFileSetPermissions;
+use sys_traits::FsMetadata;
+use sys_traits::FsOpen;
+use sys_traits::FsRemoveFile;
+use sys_traits::FsRename;
+use sys_traits::OpenOptions;
+use sys_traits::SystemRandom;
+use sys_traits::ThreadSleep;
+
+use crate::get_atomic_path;
+use crate::normalize_path;
+
+/// Canonicalizes a path which might be non-existent by going up the
+/// ancestors until it finds a directory that exists, canonicalizes
+/// that path, then adds back the remaining path components.
+///
+/// Note: When using this, you should be aware that a symlink may
+/// subsequently be created along this path by some other code.
+pub fn canonicalize_path_maybe_not_exists(
+  sys: &impl FsCanonicalize,
+  path: &Path,
+) -> std::io::Result<PathBuf> {
+  let path = normalize_path(path);
+  let mut path = path.as_path();
+  let mut names_stack = Vec::new();
+  loop {
+    match sys.fs_canonicalize(path) {
+      Ok(mut canonicalized_path) => {
+        for name in names_stack.into_iter().rev() {
+          canonicalized_path = canonicalized_path.join(name);
+        }
+        return Ok(canonicalized_path);
+      }
+      Err(err) if err.kind() == ErrorKind::NotFound => {
+        names_stack.push(match path.file_name() {
+          Some(name) => name.to_owned(),
+          None => return Err(err),
+        });
+        path = match path.parent() {
+          Some(parent) => parent,
+          None => return Err(err),
+        };
+      }
+      Err(err) => return Err(err),
+    }
+  }
+}
+
+pub fn atomic_write_file_with_retries<
+  TSys: FsCreateDirAll
+    + FsMetadata
+    + FsOpen
+    + FsRemoveFile
+    + FsRename
+    + ThreadSleep
+    + SystemRandom,
```

Comment:
I know this seems slightly out of control, but it makes using this crate much easier. A main problem atm is Rust doesn't have a way to create aliases of traits, so there's no way to just say "`deno_path_util` uses these traits under the name `DenoPathUtilSys`, so now accept `TSys: DenoPathUtilSys`". Yes, we could do `pub trait DenoPathUtilSys: FsCreateDirAll + ... {}`, but a huge negative to that is then people need to implement the `DenoPathUtilSys` crate for file system implementation and that might not always be possible if the file system implementation lives in a different crate... so we add some verbosity here for a better experience for consumers.